src/Formik.tsx

```diff
@@ -366,20 +366,33 @@ export class Formik<Values = {}, ExtraProps = {}> extends React.Component<
     value: any,
     shouldValidate: boolean = true
   ) => {
-    if (this.didMount) {
+    return new Promise((resolve, reject) => {
+      if (!this.didMount) {
+        resolve();
+      }
       // Set form field by name
       this.setState(
         prevState => ({
           ...prevState,
           values: setIn(prevState.values, field, value),
         }),
         () => {
+          const values = this.state.values;
           if (this.props.validateOnChange && shouldValidate) {
-            this.runValidations(this.state.values);
+            this.runValidations(values).then(combinedErrors => {
+              const isValid = Object.keys(combinedErrors).length === 0;
+              if (!isValid) {
+                reject(combinedErrors);
```

Comment:
Honestly, I am not really sure if those methods should return any value at all. Most of the use cases just want to wait until it's set to call different handler. For example:

```js
onClick={async value => {
  await setFieldValue(name, value)
  submitForm() // or some different handler, which then reads necessary values from the state
}
```

what would be the use case for:

```js
onClick={async value => {
  try {
    const value = await setFieldValue(name, value) // what now?
  } catch (errors) {
     console.log(errors) // all errors, why?
  }
}
```

instead of:
```js
onClick={async value => {
    console.log(formikBag)
    await setFieldValue(name, value)
    console.log(formikBag) // updated
}
```

Maybe I am missing something, will wait for other thoughts.

Reply:
I'll have a play around tomorrow morning and see if I can come up with an example. I think I have a use case in mind but I might be overthinking things

Reply:
@Woile Yeah I was overthinking it. I was worried about a stale `errors` object after the Promise resolves but that's not the case in practice. I agree with @prichodko :)